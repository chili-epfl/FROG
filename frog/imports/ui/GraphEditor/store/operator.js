// @flow
import cuid from 'cuid';
import { observable, action, computed } from 'mobx';
import { store } from './index';
import Elem from './elemClass';
import { pxToTime, timeToPx } from '../utils';
import type { AnchorT } from '../utils/path';

export default class Operator extends Elem {
  id: string;
  type: string;
  klass: string;
  @observable y: number;
  @observable title: ?string;
  @observable over: boolean;
  @observable time: number;
  @observable wasMoved: boolean = false;

  @action
  init(time: number, y: number, type: string, id: ?string, title: ?string) {
    this.time = time;
    this.y = y;
    this.id = id || cuid();
    this.type = type;
    this.klass = 'operator';
    this.title = title;
  }

  constructor(
    time: number,
    y: number,
    type: string,
    id: ?string,
    title: ?string
  ) {
    super();
    this.init(time, y, type, id, title);
  }

  @computed
  get x(): number {
    return timeToPx(this.time, 4);
  }

  @computed
  get xScaled(): number {
    return timeToPx(this.time, store.ui.scale);
  }

  @computed
  get coordsScaled(): [number, number] {
    return [this.xScaled, this.y];
  }

  @computed
  get coords(): [number, number] {
    return [this.x, this.y];
  }

  @action rename = (newname: string) => {
    this.title = newname;
    store.addHistory();
  };

  @action onOver = (): true => (this.over = true);
  @action onLeave = (): false => (this.over = false);

  @action startDragging = (e: { shiftKey: boolean }): void => {
    if (!e.shiftKey) {
      store.connectionStore.startDragging(this);
    } else {
      store.state = { mode: 'movingOperator', currentOperator: this };
    }
  };

  @action onDrag = (
    e: { shiftKey: boolean },
    { deltaX, deltaY }: { deltaX: number, deltaY: number }
  ) => {
    if (store.state.mode === 'movingOperator') {
      this.time += pxToTime(deltaX, store.ui.scale);
      this.y += deltaY;
      this.wasMoved = true;
    }
  };

  @action moveX = (deltaX: number): void => {
    this.time += pxToTime(deltaX, store.ui.scale);
    this.wasMoved = true;
  };

  @action stopDragging = (): void => {
    if (store.state.mode === 'movingOperator') {
      store.state = { mode: 'normal' };
      store.addHistory();
    } else {
      store.connectionStore.stopDragging();
    }
    store.ui.cancelScroll();
  };

  @computed
  get object(): {
    _id: string,
    time: number,
    y: number,
    type: string,
    title: ?string
  } {
    return {
      _id: this.id,
      time: this.time,
      y: this.y,
      type: this.type,
      title: this.title
    };
  }

  @computed
  get dragPointTo(): AnchorT {
    // operator has size of 60, finding midpoint
    return {
      X: this.x + 25,
      Y: this.y + 25,
      dX: -150,
      dY: 0
    };
  }

  @computed
  get dragPointFrom(): AnchorT {
    // operator has size of 60, finding midpoint
    return {
      X: this.x + 25,
      Y: this.y + 25,
      dX: 150,
      dY: 0
    };
  }

  @computed
  get dragPointToScaled(): AnchorT {
    // operator has size of 60, finding midpoint
    return {
      X: this.xScaled + 25,
      Y: this.y + 25,
      dX: -150,
      dY: 0
    };
  }

  @computed
  get dragPointFromScaled(): AnchorT {
    // operator has size of 60, finding midpoint
    return {
      X: this.xScaled + 25,
      Y: this.y + 25,
      dX: 150,
      dY: 0
    };
  }

  @action update = (newopt: $Shape<Operator>) => {
    this.time = newopt.time;
    this.y = newopt.y;
    this.title = newopt.title;
  };
}
