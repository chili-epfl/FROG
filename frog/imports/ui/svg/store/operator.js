// @flow
import cuid from 'cuid';
import { observable, action, computed } from 'mobx';
import { store } from './index';
import { pxToTime, timeToPx } from '../utils';

export default class Operator {
  id: string;
  type: string;
  @observable selected: boolean;
  @observable y: number;
  @observable over: boolean;
  @observable time: number;
  @action init(time: number, y: number, type: string, id: string) {
    this.time = time;
    this.y = y;
    this.id = id || cuid();
    this.type = type;
  }

  constructor(time: number, y: number, type: string, id: string) {
    this.init(time, y, type, id);
  }

  @computed get x(): number {
    return timeToPx(this.time, 1);
  }

  @computed get xScaled(): number {
    return timeToPx(this.time, store.scale);
  }

  @computed get coordsScaled(): [number, number] {
    return [this.xScaled, this.y];
  }

  @computed get coords(): [number, number] {
    return [this.x, this.y];
  }

  @action onClick = (): void => {
    store.unselect();
    this.selected = true;
  };

  @action onOver = (): true => this.over = true;
  @action onLeave = (): false => this.over = false;
  @computed get highlighted(): boolean {
    return this.over &&
      store.draggingFromActivity !== this &&
      store.mode === 'dragging';
  }

  @action startDragging = (e: { shiftKey: boolean }): void => {
    if (!e.shiftKey) {
      store.startDragging(this);
    }
  };

  @action onDrag = (
    e: { shiftKey: boolean },
    { deltaX, deltaY }: { deltaX: number, deltaY: number }
  ) => {
    if (!e.shiftKey) {
      store.dragging(deltaX, deltaY);
    } else {
      this.time += pxToTime(deltaX, store.scale);
      this.y += deltaY;
    }
  };

  @action stopDragging = (e: { shiftKey: boolean }): void => {
    if (!e.shiftKey) {
      store.stopDragging();
    }
    store.addHistory();
  };

  @computed get object(): {
    _id: string,
    time: number,
    y: number,
    type: string
  } {
    return {
      _id: this.id,
      time: this.time,
      y: this.y,
      type: this.type
    };
  }

  @computed get dragPoint(): [number, number] {
    return [this.xScaled + 30, this.y + 30];
  }

  @action update = (newopt: $Shape<Operator>) => {
    this.time = newopt.time;
    this.y = newopt.y;
  };
}
