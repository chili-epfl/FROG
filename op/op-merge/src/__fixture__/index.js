export const object = {
  _id: 'cjstg14jr0007zjiy8ifcox1o',
  activityData: {
    cjstg14jr0002zjiyiunjmwpp: {
      structure: 'individual',
      payload: {
        '8SRWeEbbSwKkZYy8f': { data: {} },
        Dt39sCkvuZC7Ew8iB: {
          data: {
            cjstg1gfi00003f6n732zj0x3: {
              order: 1,
              msg: 'a',
              user: 'Chen Li',
              id: 'cjstg1gfi00003f6n732zj0x3'
            }
          }
        },
        fhCWLQBNJzt8LeQqX: { data: {} },
        mxJdzWBAfLWJxGF8e: {
          data: {
            cjstg1jfi00003f6n01s6ozl2: {
              order: 1,
              msg: 'd',
              user: 'Peter',
              id: 'cjstg1jfi00003f6n01s6ozl2'
            }
          }
        }
      }
    },
    cjstg14jr0003zjiy3k6o8v0d: {
      structure: 'individual',
      payload: {
        '8SRWeEbbSwKkZYy8f': {
          data: {
            cjstg1ht300003f6ntk7p7p86: {
              order: 1,
              msg: 'b',
              user: 'Anna',
              id: 'cjstg1ht300003f6ntk7p7p86'
            }
          }
        },
        Dt39sCkvuZC7Ew8iB: { data: {} },
        fhCWLQBNJzt8LeQqX: {
          data: {
            cjstg1ija00003f6ndcozph3k: {
              order: 1,
              msg: 'c',
              user: 'Aliya',
              id: 'cjstg1ija00003f6ndcozph3k'
            }
          }
        },
        mxJdzWBAfLWJxGF8e: { data: {} }
      }
    }
  },
  socialStructure: {
    group: {
      '1': ['fhCWLQBNJzt8LeQqX', '8SRWeEbbSwKkZYy8f'],
      '2': ['Dt39sCkvuZC7Ew8iB', 'mxJdzWBAfLWJxGF8e']
    }
  },
  globalStructure: {
    studentIds: [
      'Dt39sCkvuZC7Ew8iB',
      'fhCWLQBNJzt8LeQqX',
      'mxJdzWBAfLWJxGF8e',
      '8SRWeEbbSwKkZYy8f'
    ],
    students: {
      Dt39sCkvuZC7Ew8iB: 'Chen Li',
      fhCWLQBNJzt8LeQqX: 'Aliya',
      mxJdzWBAfLWJxGF8e: 'Peter',
      '8SRWeEbbSwKkZYy8f': 'Anna'
    }
  }
};

export const object2 = {
  _id: 'cjstw7f230034t6se4h3sh7i0',
  activityData: {
    cjstw7f23002qt6seyke9d42w: {
      structure: 'individual',
      payload: {
        '8SRWeEbbSwKkZYy8f': {
          data: {
            answers: {
              '8SRWeEbbSwKkZYy8f': {
                'Drag and Drop': 1,
                Form: 2,
                'Command Line': 3,
                Map: 4
              }
            },
            group: { '8SRWeEbbSwKkZYy8f': 'Anna' },
            round: 1,
            coordinates: { x: 6.4, y: -1.6 },
            alert: false,
            completed: true,
            msg:
              'Anna ranked the interfaces in the following order: Drag and Drop, Form, Command Line, Map'
          }
        },
        Dt39sCkvuZC7Ew8iB: {
          data: {
            answers: {
              Dt39sCkvuZC7Ew8iB: {
                Map: 1,
                'Command Line': 2,
                'Drag and Drop': 3,
                Form: 4
              }
            },
            group: { Dt39sCkvuZC7Ew8iB: 'Chen Li' },
            round: 1,
            coordinates: { x: -6.4, y: -3.2 },
            alert: false,
            completed: true,
            msg:
              'Chen Li ranked the interfaces in the following order: Map, Command Line, Drag and Drop, Form'
          }
        },
        fhCWLQBNJzt8LeQqX: {
          data: {
            answers: {},
            group: {},
            round: 0,
            msg:
              'Uh Oh !! The student assigned to work with you has not completed the previous activity. If he still does not participate, you could discuss with your fellow student sitting next to you.'
          }
        },
        mxJdzWBAfLWJxGF8e: {
          data: {
            answers: {},
            group: {},
            round: 0,
            msg:
              'Uh Oh !! The student assigned to work with you has not completed the previous activity. If he still does not participate, you could discuss with your fellow student sitting next to you.'
          }
        }
      }
    },
    cjstw7f23002xt6set9dqwywf: {
      structure: 'individual',
      payload: {
        '8SRWeEbbSwKkZYy8f': {
          data: {
            answers: {},
            group: {},
            round: 0,
            msg:
              'Uh Oh !! The student assigned to work with you has not completed the previous activity. If he still does not participate, you could discuss with your fellow student sitting next to you.'
          }
        },
        Dt39sCkvuZC7Ew8iB: {
          data: {
            answers: {},
            group: {},
            round: 0,
            msg:
              'Uh Oh !! The student assigned to work with you has not completed the previous activity. If he still does not participate, you could discuss with your fellow student sitting next to you.'
          }
        },
        mxJdzWBAfLWJxGF8e: {
          data: {
            answers: {
              mxJdzWBAfLWJxGF8e: {
                Map: 1,
                'Command Line': 2,
                'Drag and Drop': 3,
                Form: 4
              }
            },
            group: { mxJdzWBAfLWJxGF8e: 'Peter' },
            round: 1,
            coordinates: { x: -6.4, y: -3.2 },
            alert: false,
            completed: true,
            msg:
              'Peter ranked the interfaces in the following order: Map, Command Line, Drag and Drop, Form'
          }
        }
      }
    }
  },
  socialStructure: {},
  globalStructure: {
    studentIds: [
      'Dt39sCkvuZC7Ew8iB',
      'fhCWLQBNJzt8LeQqX',
      'mxJdzWBAfLWJxGF8e',
      '8SRWeEbbSwKkZYy8f'
    ],
    students: {
      Dt39sCkvuZC7Ew8iB: 'Chen Li',
      fhCWLQBNJzt8LeQqX: 'Aliya',
      mxJdzWBAfLWJxGF8e: 'Peter',
      '8SRWeEbbSwKkZYy8f': 'Anna'
    }
  }
};
