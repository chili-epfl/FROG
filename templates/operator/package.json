{
  "name": "operator-template",
  "version": "1.0.0",
  "description": "",
  "main": "src/index.js",
  "scripts": {
    "start": "nps"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {},
  "devDependencies": {
    "nps": "5.9.3"
  }
}
