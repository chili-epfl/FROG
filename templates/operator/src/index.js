// @flow

import { shuffle, chunk } from 'lodash';
import { type productOperatorT } from 'frog-utils';

const meta = {
  name: 'Argue',
  shortDesc: 'Group students to argue',
  description: 'Group students with as many similar answers as possible.'
};

const config = {
  type: 'object',
  properties: {}
};

export default ({
  id: 'op-argue',
  type: 'product',
  configVersion: 1,
  config,
  meta
}: productOperatorT);
