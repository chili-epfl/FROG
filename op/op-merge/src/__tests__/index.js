import operator from '../operatorRunner';
import { object, object2 } from '../__fixture__';

test('Merge', () => {
  expect(operator({}, object)).toMatchSnapshot();
});

test('Merge when empty are not empty', () => {
  expect(operator({}, object2)).toMatchSnapshot();
});
