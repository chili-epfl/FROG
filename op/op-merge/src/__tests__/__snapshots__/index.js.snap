// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge 1`] = `
Object {
  "payload": Object {
    "8SRWeEbbSwKkZYy8f": Object {
      "data": Object {
        "cjstg1ht300003f6ntk7p7p86": Object {
          "id": "cjstg1ht300003f6ntk7p7p86",
          "msg": "b",
          "order": 1,
          "user": "Anna",
        },
      },
    },
    "Dt39sCkvuZC7Ew8iB": Object {
      "data": Object {
        "cjstg1gfi00003f6n732zj0x3": Object {
          "id": "cjstg1gfi00003f6n732zj0x3",
          "msg": "a",
          "order": 1,
          "user": "Chen Li",
        },
      },
    },
    "fhCWLQBNJzt8LeQqX": Object {
      "data": Object {
        "cjstg1ija00003f6ndcozph3k": Object {
          "id": "cjstg1ija00003f6ndcozph3k",
          "msg": "c",
          "order": 1,
          "user": "Aliya",
        },
      },
    },
    "mxJdzWBAfLWJxGF8e": Object {
      "data": Object {
        "cjstg1jfi00003f6n01s6ozl2": Object {
          "id": "cjstg1jfi00003f6n01s6ozl2",
          "msg": "d",
          "order": 1,
          "user": "Peter",
        },
      },
    },
  },
  "structure": "individual",
}
`;

exports[`Merge when empty are not empty 1`] = `
Object {
  "payload": Object {
    "8SRWeEbbSwKkZYy8f": Object {
      "data": Object {
        "alert": false,
        "answers": Object {
          "8SRWeEbbSwKkZYy8f": Object {
            "Command Line": 3,
            "Drag and Drop": 1,
            "Form": 2,
            "Map": 4,
          },
        },
        "completed": true,
        "coordinates": Object {
          "x": 6.4,
          "y": -1.6,
        },
        "group": Object {
          "8SRWeEbbSwKkZYy8f": "Anna",
        },
        "msg": "Anna ranked the interfaces in the following order: Drag and Drop, Form, Command Line, Map",
        "round": 1,
      },
    },
    "Dt39sCkvuZC7Ew8iB": Object {
      "data": Object {
        "alert": false,
        "answers": Object {
          "Dt39sCkvuZC7Ew8iB": Object {
            "Command Line": 2,
            "Drag and Drop": 3,
            "Form": 4,
            "Map": 1,
          },
        },
        "completed": true,
        "coordinates": Object {
          "x": -6.4,
          "y": -3.2,
        },
        "group": Object {
          "Dt39sCkvuZC7Ew8iB": "Chen Li",
        },
        "msg": "Chen Li ranked the interfaces in the following order: Map, Command Line, Drag and Drop, Form",
        "round": 1,
      },
    },
    "fhCWLQBNJzt8LeQqX": Object {
      "data": Object {
        "answers": Object {},
        "group": Object {},
        "msg": "Uh Oh !! The student assigned to work with you has not completed the previous activity. If he still does not participate, you could discuss with your fellow student sitting next to you.",
        "round": 0,
      },
    },
    "mxJdzWBAfLWJxGF8e": Object {
      "data": Object {
        "alert": false,
        "answers": Object {
          "mxJdzWBAfLWJxGF8e": Object {
            "Command Line": 2,
            "Drag and Drop": 3,
            "Form": 4,
            "Map": 1,
          },
        },
        "completed": true,
        "coordinates": Object {
          "x": -6.4,
          "y": -3.2,
        },
        "group": Object {
          "mxJdzWBAfLWJxGF8e": "Peter",
        },
        "msg": "Peter ranked the interfaces in the following order: Map, Command Line, Drag and Drop, Form",
        "round": 1,
      },
    },
  },
  "structure": "individual",
}
`;
